text/0000-new-test-waiters.md

```diff
@@ -85,21 +85,114 @@ export default class Friendz extends Component {
 
 In the above example, a new test waiter is built that is identified via a `name` string passed into the `buildWaiter` function. This allows the waiter to be identifiable, and that name is ultimately used with test isolation validation to help developers narrow down problems in their tests.
 
-The API used to signal whether an asynchronous operation has begun and ultimately ended is through the paired calls of `beginAsync` and `endAsync`: begin to denote the start of the asynchronous operation, and end to denote the end. Unique instances of async operations are identified using a `token` returned from `beginAsync`, which is subsequently provided to the `endAsync` call.
+## New Test Waiters Design
+
+The new test waiters addon is built using low-level primitives that are complimented with some convenience utilities.
+
+### `IWaiter`
+
+At its core, the addon uses an `IWaiter` interface defined as follows:
+
+```ts
+export type WaiterName = string;
+export type Token = unknown;
+
+export interface IWaiter {
+  name: WaiterName;
+  waitUntil(): boolean;
+  debugInfo(): ITestWaiterDebugInfo[];
+}
+```
+
+This allows for maximum flexibility when creating your own waiter implementations.
+
+### `ITestWaiter`
+
+The `IWaiter` interface is built upon to create a more specific interface for a test waiter, `ITestWaiter`:
+
+```ts
+export interface ITestWaiter<T = Token> extends IWaiter {
+  beginAsync(token?: T, label?: string): T;
+  endAsync(token: T): void;
+  reset(): void;
+}
+```
+
+This interface is used for the two internal concrete types: `TestWaiter` and `NoopTestWaiter`. These types for the basis for the addon, and will likely satisfy the majority of use cases.
+
+The most common practice is to import and invoke the `buildWaiter` function to create a new test waiter. The recommendation is to do so at the module level, which allows a single waiter to be created per type. A single waiter is then usable across multiple instances.
+
+```ts
+function buildWaiter(name: string): ITestWaiter
+```
+
+In anything but a production build, this function will return a `TestWaiter` instance. When in production mode, a `NoopTestWaiter` will be returned, which as the name suggests is effectively a noop class. Since test waiters are intended to be called from application or addon code, but are only required to be _active_ when in tests, we use the `NoopTestWaiter` to noop the class' methods when run in production. This has a negligible impact on performance.
```

Comment:
I think we can avoid mentioning NoopTestWaiter specifically, but instead mention that in production (not testing) builds that the `TestWaiter` class is "inert" and optimized for "speed".

Reply:
I see no harm in mentioning it, just so people understand how it works slightly more than hand waving. Any issues with my leaving it in?

Reply:
The issue is that the _how_ we make it inert is not really important (and definitely isn't public api). I think making it clear what our commitment is (that it is effectively zero cost in production), and that "for example, you could imagine that we use an alternative class implementation in production that makes all operations a no op".

Reply:
Sure. I can go either way, so will remove it.